### PR TITLE
Allow timed jobs' intervals to be overriden by config (SC-308)

### PR DESCRIPTION
**Pull request by renanrodrigo**
User can now set the desired run_interval for jobs in the `uaclient.conf` file, as in:
```
ua_config:
    <job_name>_timer: <run_interval_value>
```
It has the config for the messaging and status jobs.

## Desired commit type::
 - [ ] Squash and merge: Using the "Proposed Commit Message"
 - [ ] Create a merge commit and use "Proposed Commit Message"
 - [x] Rebase and merge, no merge commit, rely only on existing commit messages

## Checklist:
 - [x] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any integration tests accordingly
 - [ ] I have updated or added any documentation accordingly

### CODE REVIEW

File: lib/timer.py
Comment (lucasmoura) [inline lib/timer.py:50]: Why do we need a Timer class ?

File: uaclient/config.py
Comment (lucasmoura) [inline uaclient/config.py:351]: I think the idea here is to allow user to set the jobs time configurations through `ua config set`
We will probably need to move that logic under the `ua_config` key

File: lib/timer.py
Comment (renanrodrigo) [reply lib/timer.py:51]: We do not - I thought it would be a more organized and readable way to encapsulate the jobs and the running functionality. and since Jobs were turned into classes, I created the Timer one as well.

File: uaclient/config.py
Comment (renanrodrigo) [reply uaclient/config.py:351]: Sure, can take a look into that.
Are you thinking about something like `ua cofig set timer <job name> run_interval_seconds <value>`?

Comment (renanrodrigo) [reply uaclient/config.py:351]: Speaking about that, if we do not want to configure anything but the intervals, is there any major downside in going
```
timer:
    <job_name>: <interval_value>
```
for config?

File: lib/timer.py
Comment (blackboxsw) [inline lib/timer.py:9]: do we still need the noqa comment?

File: uaclient/config.py
Comment (blackboxsw) [reply uaclient/config.py:351]: @lucasmoura @renanrodrigo I vote to avoid adding a `ua config set <timer> <runinterval>` for this release. I'd expect that custom setting timer run frequency to something slower than what our defaults are is not a highly used config operation that we expect lots of users to "want". But we can discuss this with @mitechie as to whether this use case is one that begs the CLI "sugar" for ease of set/unset.

File: lib/timer.py
Comment (blackboxsw) [reply lib/timer.py:51]: I like have you've ecapsulated the TimedJob.run definition, but I'm a little uncertain if this is overengineering the "Timer" class at the moment since we are only really talking about an ordered dict and a single function to run_jobs. I also have a general aversion to __init__ methods in classes which have a lot of side-effects (like initializing UAConfig and multiple TimedJob instances. Are there other compelling reasons to make this a class?

**Review by blackboxsw — CHANGES_REQUESTED** [review]
@renanrodrigo general approach looks good. +1 on the TimedJob instrumentation, have a couple questions about the real motivation at this PR for a Timer class, but the only significant thing I really want to avoid at the moment in this PR is CLI changes to `ua config`.

Per your questions:
Opening it as a draft, for I would wait until #1761 is merged, and then I can add the metering job here as well - imho it will be easier for me to rebase this than merging this and then rebasing the metering PR.
   - yeah you can rebase off of @lucasmoura's branch as a base. I've approved that so I imagine he can land it early in the day when CI is sorted.
   - Feel free to open a doc PR separately so we can add a doc for the full timer and both existing jobs etc. will be easier to review docs sepearately
   - Should I write any integration tests for this?   Please do extend Lucas' existing integraiton test "Run timer script on an attached machine" you might be able to just set/append to uaclient.conf and allow the timer to trigger and check the next_run value in /var/lib/ubuntu-advantage/jobs-status.json to ensure override was applied.

File: lib/timer.py
Comment (renanrodrigo) [reply lib/timer.py:51]: No other compelling. Will change it back to a function.

File: lib/timer.py
Comment (renanrodrigo) [reply lib/timer.py:55]: no class anymore so no reason for this

File: lib/timer.py
Comment (renanrodrigo) [reply lib/timer.py:9]: nope (:

File: uaclient/config.py
Comment (renanrodrigo) [reply uaclient/config.py:351]: Maybe we could do this iteration using only the config file, and implement the commands in another PR afterwards

**Comment by blackboxsw** [comment]
Per discussion at standup today, we **want** to support `ua config set timer ...`  whether you want to do that in this branch or a separate branch let's iron that out in the spec file. I'll add a section now for you and Lucas to discuss tomorrow.

**Comment by renanrodrigo** [comment]
I can do it in this one, no problem.

File: uaclient.conf
Comment (lucasmoura) [inline uaclient.conf:15]: One idea that @mitechie proposed is to ship the config with the default values already in place.
If we do that, we can remove the constants we are using on the timer module

File: lib/timer.py
Comment (renanrodrigo) [reply lib/timer.py:58]: No. As the configuration has all the job timers set to null by default, `getattr` will return `None` and the default value will not be returned.

File: uaclient/tests/test_ua_timer.py
Comment (renanrodrigo) [reply uaclient/tests/test_ua_timer.py:94]: 1. As this is a unit test, a generic entry here  tests only the functionality of this piece of code, without depending on the real environment around it
2. Follows from 1: when any new job is added, there is no side effect, and no need to change the test because it is one more item to iterate over the same logic

File: uaclient.conf
Comment (renanrodrigo) [reply uaclient.conf:15]: Ack, but then, what happens if the user sets the value to `null`?
We would need to change some hereandtheres to make sure it will not surface the None value, as we would not have the constants.

So, imho, it is ok to put the defaults here, but keeping them in the file is safer.

Comment (renanrodrigo) [reply uaclient.conf:15]: As discussed with @mitechie, we can define the same default values in the config file, but keep those in the constants as a fallback value for any invalid or missing value in the config file.

File: lib/timer.py
Comment (renanrodrigo) [reply lib/timer.py:58]: Config will not be set to null in the end, but to some integer value. I will change this to ignore wrong attribute values and log occurrences, using the default when it happens,

File: uaclient/config.py
Comment (renanrodrigo) [reply uaclient/config.py:795]: Will not fail here anymore, just log an error if the validation goes wrong

File: uaclient/tests/test_ua_timer.py
Comment (lucasmoura) [reply uaclient/tests/test_ua_timer.py:94]: I get the idea, but I do like the idea that for every job we run, we will need to modify that test. It should at least cover that we are properly configuring the time for each job correctly. But let me know if you think this is an overkill, as I don't have a strong opinion on that

Comment (renanrodrigo) [reply uaclient/tests/test_ua_timer.py:94]: I see. I think that to test the jobs themselves, with the real values, a behave test would make more sense  - can try and orchestrate a way of checking if next_run is being calculated appropriately.

File: lib/timer.py
Comment (lucasmoura) [inline lib/timer.py:63]: I think we can drop this `the` and replace `Found` for `found`

Comment (lucasmoura) [reply lib/timer.py:63]: Also, can we test that scenario on the behave test as well ? See if we are logging this part if an invalid config is provided

File: uaclient/cli.py
Comment (lucasmoura) [inline uaclient/cli.py:684]: I think we should have a message here saying why we raised the value error.
Furthermore, maybe it is a good to idea to also have this validation on the behave test

File: uaclient/tests/test_ua_timer.py
Comment (lucasmoura) [reply uaclient/tests/test_ua_timer.py:86]: I think this be harder to check on the behave tests, but if you think there is an easy way to do that, go for it

File: uaclient/config.py
Comment (lucasmoura) [inline uaclient/config.py:805]: I think we can abort the method if we detect this, similar to what we do with the proxy config

File: uaclient/tests/test_ua_timer.py
Comment (lucasmoura) [inline uaclient/tests/test_ua_timer.py:67]: I think we can turn this into a mock and assert that the mock was never called, just to be sure that the reason the call failed is because of the timer and not the function itself

File: uaclient/config.py
Comment (renanrodrigo) [reply uaclient/config.py:805]: I remember @mitechie said to avoid failing when dealing with the configuration files. CLI fails and does not set the value, but processing the config should not fail.

Comment (lucasmoura) [reply uaclient/config.py:805]: But this is the user manually running `ua refresh config`. In that situation, I think we should fail early to alert the user about the mistake.

File: uaclient/tests/test_config.py
Comment (lucasmoura) [inline uaclient/tests/test_config.py:1423]: Can we check the exception message here to make sure we failing due to the expected exception ?

**Review by lucasmoura — APPROVED** [review]
LGTM, let's just wait for CI before landing it

**Review by blackboxsw — APPROVED** [review]
LGTM. Let's make sure we circle back to the systemd timer being run every < 30 mins to support the gcp auto-attach frequency in a separate PR.

**Pull request merged**
